### PR TITLE
Meta: Fix feature details in Options page

### PR DESCRIPTION
**Pull request by sharkykh**
I'm still fairly new to TypeScript, so hopefully this is okay.

It only uses the first match it finds, so for example, on [default-to-rich-diff.tsx](https://github.com/sindresorhus/refined-github/blob/master/source/features/default-to-rich-diff.tsx), `disabled` is matched first, and so `description` is `undefined`.
>

### CODE REVIEW

File: webpack.config.ts [hunk 1]
```diff
@@ -11,13 +11,28 @@ import ForkTsCheckerWebpackPlugin from 'fork-ts-checker-webpack-plugin';
...
-	return feature as unknown as FeatureInfo;
+	const feature: FeatureInfo = {
+		name,
+		description: ''
+	};
+
+	// Iterate over all the matches from the file
+	let match: RegExpExecArray | null;
+	while ((match = regex.exec(content)) !== null) {
+		const {field, value} = match.groups as { field: typeof fields[number]; value: string };
```
Comment (sharkykh) [inline webpack.config.ts]: `typeof fields[number]` explanation: https://stackoverflow.com/a/45486495/7597273

**Comment by sharkykh** [comment]
> My ideal fix would be to move this meta data to a comment in the file header, it's easier and safer to parse as plain text. e.g.
> 
> ```js
> /*
> Insert collapsible content when writing comments (via `<summary>`)
> 
> https://user-images.githubusercontent.com/1402241/53678019-0c721280-3cf4-11e9-9c24-4d11a697f67c.png
> */
> ```

How would you tell `disabled`, `screenshot` and `description` apart in this format?

**Comment by bfred-it** [comment]
`description` is first line
`screenshot` is the URL
`disabled` would still be in the `features.add` field

File: webpack.config.ts [hunk 2]
```diff
@@ -11,13 +11,28 @@ import ForkTsCheckerWebpackPlugin from 'fork-ts-checker-webpack-plugin';
 
 function parseFeatureDetails(name: string): FeatureInfo {
 	const content = readFileSync(`source/features/${name}.tsx`, {encoding: 'utf-8'});
-	const rawRegex = ['description', 'screenshot', 'disabled']
-		.map(field => `\n\t${field}: '(?<${field}>[^\\n]+)'`) // Named group regex
-		.join('|');
+	const fields = ['description', 'screenshot', 'disabled'] as const;
```
[the remark this answers is not in the document]
Comment (bfred-it) [reply webpack.config.ts]: I mean this change _instead of **all**_ the rest. The order was the problem, so only the order needs to be changed (and we can keep the error below)

[the remark this answers is not in the document]
Comment (sharkykh) [reply webpack.config.ts]: I'm pretty sure I tried that. But as I said above, this code did not work as intended/expected.

Only the first match of the RegExp pattern is being used. The others are discarded.
This is a "limitation" of JavaScript (that is, if you want to use `match.groups` to address the groups by name)

The order of the fields **does not** have an effect on which field is matched first, as that depends on the order in which the fields are defined in the content of the feature source.

Maybe we're just not on the same page - do you mean literally "all the rest"? Like, without the `while` loop over the matches ([as is suggested on MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec#Finding_successive_matches))?

**Comment by jerone** [comment]
[**@bfred-it**](https://github.com/bfred-it) commented on [Jul 10, 2019, 9:16 AM GMT+2](https://github.com/sindresorhus/refined-github/pull/2219#issuecomment-509940787 "2019-07-10T07:16:43Z - Replied by Github Reply Comments"):
> `description` is first line  
> `screenshot` is the URL  
> `disabled` would still be in the `features.add` field

Or we implement something similar to userscript metadata block:
```js
// ==UserScript==
// @key value
// ==/UserScript==
```

Information about this:
* https://wiki.greasespot.net/Metadata_Block
* https://www.tampermonkey.net/documentation.php
* https://greasyfork.org/en/help/meta-keys

We probably only need a few right now:
```js
// ==UserScript==
// @description		Insert collapsible content when writing comments (via `<summary>`)
// @screenshotUrl	https://user-images.githubusercontent.com/1402241/53678019-0c721280-3cf4-11e9-9c24-4d11a697f67c.png
// @author			@bfred-it
// @author			@jerone
// ==/UserScript==
```

Later we can extend that with `@version`, translations (`@description:XX-YY`), on which page to load (`@include`, `@exclude`), an icon (`@icon`, `@iconURL`, `@defaulticon`), etc...

**Comment by bfred-it** [comment]
The point of using a header comment is to avoid having parsers for it. Currently Description and screenshot is all we need, and there's not much needed to parse it. Even if we wanted to add authors (which we don't, because we have git history) we'd just append `@bfred-it @jerone` which is equally easy to parse.

File: webpack.config.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (sharkykh) [reply webpack.config.ts]: ping @bfred-it, in case you missed my reply

[the remark this answers is not in the document]
Comment (bfred-it) [reply webpack.config.ts]: Indeed I had missed it. Fixed in f22055c

File: webpack.config.ts [hunk 3]
```diff
@@ -17,7 +17,19 @@ function parseFeatureDetails(name: string): FeatureInfo {
...
+			const validValue = value.trim().replace(/\\'/g, '’'); // Catch trailing spaces and incorrect apostrophes
+			if (value !== validValue) {
+				throw new Error(`
+Invalid characters found in \`${name}\`’s ${field} field. It’s
+${field}: '${value}'
+
+but it should be
+
+${field}: '${validValue}'
+`);
```
Comment (bfred-it) [inline webpack.config.ts]: Error examples:

```
                                throw new Error(`
          ^
Error: 
Invalid characters found in `swap-branches-on-compare`’s disabled field. It’s
disabled: ' https://user-images.githubusercontent.com/1402241/51669708-9a712400-1ff7-11e9-913a-ac1ea1050975.png '

but it should be

disabled: 'https://user-images.githubusercontent.com/1402241/51669708-9a712400-1ff7-11e9-913a-ac1ea1050975.png'

```

```
                                throw new Error(`
          ^
Error: 
Invalid characters found in `swap-branches-on-compare`’s description field. It’s
description: ' Swap head and base branches\' in the branch compare view'

but it should be

description: 'Swap head and base branches’ in the branch compare view'
```

Comment (bfred-it) [reply webpack.config.ts]: Shortened to

```diff
Invalid characters found in `swap-branches-on-compare`. Apply this patch:

- description: ' Swap head and base branches\' in the branch compare view'
+ description: 'Swap head and base branches’ in the branch compare view'
```

**Comment by sharkykh** [comment]
@bfred-it I wish I had thought about the solution you just pushed. Much cleaner, much more simple. 👍

----
Continuing the discussion about the header comment:

I think having a comment block with plain text (I mean without any type of `key: value` structure), while fairly easy to parse (compared to using regular expressions), could also be a bit fragile, so you might forget a new line, and now the feature has no screenshot even though it's in the file.

Furthermore, 
> `description` is first line
> `screenshot` is the URL (edit: specifically a URL on the second line)
> `disabled` would still be in the `features.add` field

You still need `disabled` for the options page (for `__featuresInfo__`).
If you leave it as a `features.add` field, you still need to parse the file for it... Right?

**Comment by bfred-it** [comment]
Merging this because it's a bugfix. We can continue the discussion for any possible future changes

**Comment by bfred-it** [comment]
>  If you leave it as a `features.add` field, you still need to parse the file for it... Right?

Right. The main reason for the move would be that **description and screenshot are never part of the runtime**. Heck technically `disabled` should instead just exclude the file from webpack, but keeping it there as a property is:
- easy to exclude at runtime
- easy to implement



> a comment block with plain text ... could also be a bit fragile

Indeed it is, so is the current method of trying to parse JS with a regex. I don't really want to load up Acorn just for this, so here were are with some simple regexes and some `throw new Error`s.

I think parsing that block is relatively easy:

```js
const [, meta] = /^[/][*]([^]*)[*][/]/.exec(fileContent);
const [description, screenshot] = meta.trim().split('\n').filter(Boolean);
if(screenshot && /^https:/.test(screenshot)) {
	throw new Error(`Screenshot invalid: ${screenshot}`)
}
```

**Pull request merged**